prio/rand.c

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2018, Henry Corrigan-Gibbs
+ * 
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. 
+ */
+
+#include <limits.h>
+#include <nss/nss.h>
+#include <nss/pk11pub.h>
+#include <nspr/prinit.h>
+#include <mprio.h>
+
+#include "debug.h"
+#include "rand.h"
+#include "util.h"
+
+#define CHUNK_SIZE 8192
+
+SECStatus
+rand_init (void)
+{
+  SECStatus rv = NSS_NoDB_Init (NULL);
+  if (rv != SECSuccess) 
+    return SECFailure;
+
+  return SECSuccess;
+}
+
+static SECStatus 
+rand_bytes_internal (void *user_data, unsigned char *out, size_t n_bytes)
+{
+  // No pointer should ever be passed in.
+  if (user_data != NULL)
+    return SECFailure;
+  if (!NSS_IsInitialized ()) {
+    PRIO_DEBUG ("NSS not initialized. Call rand_init() first.");
+    return SECFailure;
+  }
+
+  SECStatus rv;
+
+  int to_go = n_bytes;
+  unsigned char *cp = out;
+  while (to_go) {
+    int to_gen = MIN (CHUNK_SIZE, to_go);
+    if ((rv = PK11_GenerateRandom (cp, to_gen)) != SECSuccess) 
+    {
+      PRIO_DEBUG ("Error calling PK11_GenerateRandom");
+      return SECFailure;
+    }
+
+    cp += CHUNK_SIZE;
+    to_go -= to_gen;
+  }
+
+  return rv;
+}
+
+SECStatus 
+rand_bytes (unsigned char *out, size_t n_bytes)
+{
+  return rand_bytes_internal (NULL, out, n_bytes);
+}
+
+SECStatus
+rand_int (mp_int *out, const mp_int *max)
+{
+  return rand_int_rng (out, max, &rand_bytes_internal, NULL);
+}
+
+SECStatus
+rand_int_rng (mp_int *out, const mp_int *max, 
+    RandBytesFunc rng_func, void *user_data)
+{
+  SECStatus rv = SECSuccess;
```

Comment:
Generally, move these `rv` down to first use.

Reply:
Done.

Reply:
Actually, one question about this: Do you want me to try to do this throughout libprio? For consistency, I have been placing the `rv` initialization at the top of the function just to keep it out of the way of the meaningful/non-repetitive code. If it's worthwhile, I can go back through all of the code and move the `rv`s down, but I'm not sure if this will improve readability.

Reply:
Okay, if it's acceptable, I will leave these as is.